config.go

```diff
@@ -0,0 +1,142 @@
+package analytics
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/segmentio/backo-go"
+	"github.com/xtgo/uuid"
+)
+
+// Instances of this type carry the different configuration options that may
+// be set when instantiating a client.
+//
+// Each field's zero-value is either meaningful or interpreted as using the
+// default value defined by the library.
+type Config struct {
+
+	// The endpoint to which the client connect and send their messages, set to
+	// `DefaultEndpoint` by default.
+	Endpoint string
+
+	// The flushing interval of the client. Messages will be sent when they've
+	// been queued up to the maximum batch size or when the flushing interval
+	// timer triggers.
+	Interval time.Duration
+
+	// The HTTP transport used by the client, this allows an application to
+	// redefine how requests are being sent at the HTTP level (for example,
+	// to change the connection pooling policy).
+	// If none is specified the client uses `http.DefaultTransport`.
+	Transport http.Transport
+
+	// The logger used by the client to output info or error messages when that
+	// are generated by background operations.
+	// If none is specified the client uses a standard logger that outputs to
+	// `os.Stderr`.
+	Logger Logger
+
+	// The maximum number of messages that will be sent in one API call.
+	// Messages will be sent when they've been queued up to the maximum batch
+	// size or when the flushing interval timer triggers.
+	// Note that the API will still enfore a 500KB limit on each HTTP request
+	// which is independant from the number of embedded messages.
+	BatchSize int
+
+	// When set to true the client will send more frequent and detailed messages
+	// to its logger.
+	Verbose bool
+
+	// A function called by the client to generate unique message identifiers.
+	// The client uses a UUID generator if none is provided.
+	UID func() string
+
+	// A function called by the client to get the current time, `time.Now` is
+	// used by default.
+	Now func() time.Time
+
+	// The retry policy used by the client to resend requests that have failed.
+	// The function is called with how many times the operation has been retried
+	// and is expected to return how long the client should wait before trying
+	// again.
+	// If not set the client will fallback to use a default retry policy.
+	RetryAfter func(int) time.Duration
```

Comment:
I think that for the 99.99% cases application don't need to worry about retries and that's kind of why they'd use this library in the first place, they need something that already handle failure cases.
Not to mention that the error is supposed to be a one-in-a-billion times event, especially for server-to-server communication where connections are all wired and high quality (which this library is designed for).

I feel like ultimately we may want to figure out a way to report messages that could not be sent at all so the application can figure out what they need to do with it but I doubt this will have